Sources/HAP/Server/Device.swift

```
@@ -144,15 +230,107 @@ public class Device {
             accessory.device = self
-            accessory.aid = idGenerator.next()!
+            if (accessory.aid == 0) {
+                
+                // Check to see if the aid has been stored in the configuration data
+                
+                let dsuid = accessory.deviceSpecificUniqueId
+                if let aid = configuration.aidForAccessoryUniqueId[dsuid] {
+                    accessory.aid = aid
+                }
+            }
         }
+
+        for accessory in newAccessories {
+            // Verify that the aid is indeed unique
+            if accessory.aid != 0,
+               !isUniqueAID(accessory.aid, ignoring: accessory) {
+                logger.info("Accessory \(accessory.info.name.value ?? "unknown") has a duplicate accessory ID \(accessory.aid), replacing with fresh IID")
+                accessory.aid = 0
+            }
+            if (accessory.aid == 0) {
+                // Obtain a new aid, which has not already been used
+                repeat {
+                    accessory.aid = configuration.nextAID
+                } while (!isUniqueAID(accessory.aid, ignoring: accessory))
+                configurationUpdated = true
+
+                // Store the aid in the configuration data
+                
+                let dsuid = accessory.deviceSpecificUniqueId
+                configuration.aidForAccessoryUniqueId[dsuid] = accessory.aid
+            }
+        }
+        
+        // write configuration data to persist updated aid's and notify listeners
+        
+        if configurationUpdated {
+            updatedConfiguration()
+        }
+
+        
+    }
+    
+    public func removeAccessories(_ unwantedAccessories: [Accessory]) {
+        
+        var configurationUpdated = false
+        
+        for accessory in unwantedAccessories {
+            // Ensure the initial accessory is not removed, and that the accessory is in the list
+            if accessory.aid != 1 {
```

Why shouldn't we remove the initial accessory? Also; we shouldn't silently ignore that, but trap as it is invalid API usage. Thus use something like `preconditionFailure()`.

---

According to the spec "2.5.3.3 Primary HAP Accessory Object", the first accessory exposed by a Bridge has to be the bridge itself, so cannot be removed. The called should create a new Device instead. I've modified the func to use precondition()